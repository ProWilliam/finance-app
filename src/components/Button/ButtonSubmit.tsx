import React, { useState } from 'react';
import { TouchableOpacity , Text, StyleProp, ViewStyle } from 'react-native';

// Hooks
import useAppNavigation from '../../hooks/useAppNavigation';

// Styles and Types
import styles from './ButtonSubmit.styles';
import { ButtonSubmitProps } from '../../types/components/Button.types';

const ButtonSubmit : React.FC<ButtonSubmitProps>= ({ title, color, navigationRoot, press}) => {

  const [route, setRoute] = useState('');
  const { appNavigation } = useAppNavigation();
  
  appNavigation(route)

  const butonStyles: StyleProp<ViewStyle>[]  = [
    styles.button,
    color == 'send' && styles.colorPrimaryButton,
    color == 'edit' && styles.colorSecondaryButton,
    color == 'deleted' && styles.colorTertiary,
  ]

  const actionPress = () => {
    if(press){
      press();
    }else if(navigationRoot){
      setRoute(navigationRoot)
    }
  };
 
  return (
    <TouchableOpacity 
      onPress={() => actionPress()}
      style={butonStyles}
    >
      <Text style={styles.text}>
        { title }
      </Text>
    </TouchableOpacity>
  )
}

export default ButtonSubmit;