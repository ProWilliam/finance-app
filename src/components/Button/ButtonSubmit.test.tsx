import React from 'react';
import { Text, TouchableOpacity } from 'react-native';
import renderer, { act } from 'react-test-renderer';
import ButtonSubmit from './ButtonSubmit';

// Mock useAppNavigation hook
jest.mock('../../hooks/useAppNavigation', () => ({
  __esModule: true,
  default: () => ({
    appNavigation: jest.fn(),
  }),
}));

describe('<ButtonSubmit />', () => {
  it('renders correctly with title', () => {
    const component = renderer.create(<ButtonSubmit title="Submit" color="edit" />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should apply the correct styles based on color prop', () => {
    let tree = renderer.create(<ButtonSubmit title="Submit" color="send" />).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(<ButtonSubmit title="Edit" color="edit" />).toJSON();
    expect(tree).toMatchSnapshot();

    tree = renderer.create(<ButtonSubmit title="Delete" color="deleted" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls press function when provided', () => {
    const onPressMock = jest.fn();
    const component = renderer.create(<ButtonSubmit title="Submit" color="edit" press={onPressMock} />);

    act(() => {
      component.root.findByType(TouchableOpacity).props.onPress();
    });
    
    expect(onPressMock).toHaveBeenCalled();
  });

  it('navigates to specified route when navigationRoot prop is provided', () => {
    const component = renderer.create(<ButtonSubmit title="Submit" color="edit" navigationRoot="Home" />);
    const touchable = component.root.findByType(TouchableOpacity);

    act(() => {
      touchable.props.onPress();
    });

    expect(component.root.findByType(Text).props.children).toBe('Submit');
  });

});