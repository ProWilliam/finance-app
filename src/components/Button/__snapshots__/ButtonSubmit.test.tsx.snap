// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonSubmit /> renders correctly with title 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#eaeaea",
      "borderRadius": 10,
      "height": 60,
      "justifyContent": "center",
      "marginVertical": 10,
      "opacity": 1,
      "paddingHorizontal": 15,
      "paddingVertical": 10,
      "width": "80%",
    }
  }
>
  <Text
    style={
      {
        "color": "#1E2126",
        "fontSize": 14,
        "fontWeight": "bold",
        "padding": 10,
      }
    }
  >
    Submit
  </Text>
</View>
`;

exports[`<ButtonSubmit /> should apply the correct styles based on color prop 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFDD00",
      "borderRadius": 10,
      "height": 60,
      "justifyContent": "center",
      "marginVertical": 10,
      "opacity": 1,
      "paddingHorizontal": 15,
      "paddingVertical": 10,
      "width": "80%",
    }
  }
>
  <Text
    style={
      {
        "color": "#1E2126",
        "fontSize": 14,
        "fontWeight": "bold",
        "padding": 10,
      }
    }
  >
    Submit
  </Text>
</View>
`;

exports[`<ButtonSubmit /> should apply the correct styles based on color prop 2`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#eaeaea",
      "borderRadius": 10,
      "height": 60,
      "justifyContent": "center",
      "marginVertical": 10,
      "opacity": 1,
      "paddingHorizontal": 15,
      "paddingVertical": 10,
      "width": "80%",
    }
  }
>
  <Text
    style={
      {
        "color": "#1E2126",
        "fontSize": 14,
        "fontWeight": "bold",
        "padding": 10,
      }
    }
  >
    Edit
  </Text>
</View>
`;

exports[`<ButtonSubmit /> should apply the correct styles based on color prop 3`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#F70000",
      "borderRadius": 10,
      "height": 60,
      "justifyContent": "center",
      "marginVertical": 10,
      "opacity": 1,
      "paddingHorizontal": 15,
      "paddingVertical": 10,
      "width": "80%",
    }
  }
>
  <Text
    style={
      {
        "color": "#1E2126",
        "fontSize": 14,
        "fontWeight": "bold",
        "padding": 10,
      }
    }
  >
    Delete
  </Text>
</View>
`;
